### PR TITLE
[dart-dio-next] Adds an option for using the dio_http package

### PR DESCRIPTION
**Pull request by josh-burton**
Adds a new generator option to change the dio package.

Options are dio or dio_http.

Relates to #10305

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@jaumard @josh-burton @amondnet @sbu-WBT @kuhnroyal @agilob @ahmednfwela

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache [hunk 1]
```diff
@@ -4,8 +4,8 @@ import 'dart:typed_data';
 
 import 'package:built_collection/built_collection.dart';
 import 'package:built_value/serializer.dart';
-import 'package:dio/dio.dart';
-import 'package:dio/src/parameter.dart';
+import '{{dioImport}}';
+import 'package:{{dioLibrary}}/src/parameter.dart';
```
Comment (ahmednfwela) [inline modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: won't this be included automatically if we import dio ?

Comment (josh-burton) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: No unfortunately it's not exported from dio.dart.

Comment (jaumard) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: is it something we need? If it's not exposed it's probably shouldn't be used.

Comment (kuhnroyal) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: It is something I actually added to dio for the 4.0 release and I forgot to export that file :(

Comment (jaumard) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: ha ha ok :) never mind then ^^ it's just I've check where it's used and kind of didn't understand the point of that :)

Comment (kuhnroyal) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: But that is something that can be fixed in `dio_http`.
I created https://github.com/flutterchina/dio/pull/1177 for this a while back.

Comment (ahmednfwela) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: maybe we can add a flag like `useDioHttp` which will be `useOriginalDio` and check it before importing `import 'package:{{dioLibrary}}/src/parameter.dart';`

**Comment by kuhnroyal** [comment]
Can you create a sample config in `bin/configs` and add the sample to the `samples.dart-2.13` in the `pom.xml`. Then we can run it automatically in CI.

File: modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache (hunk 1 above)
Comment (kuhnroyal) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: I think we currently need this import for both options.

Comment (josh-burton) [reply modules/openapi-generator/src/main/resources/dart/libraries/dio/serialization/built_value/api_util.mustache:8]: if we can get it exported in the dio_http package we can add some conditional logic to remove the import but for now it needs to be there.

**Comment by josh-burton** [comment]
> Can you create a sample config in `bin/configs` and add the sample to the `samples.dart-2.13` in the `pom.xml`. Then we can run it automatically in CI.

@kuhnroyal I've added the sample config, a pom.xml and added that pom to the project pom.xml. I'm not sure how the _tests sample gets generated though (e.g. `samples/openapi3/client/petstore/dart-dio-next/petstore_client_lib_fake_tests`)

**Comment by kuhnroyal** [comment]
Those are written by hand. Probably don't add those for now.

File: samples/openapi3/client/petstore/dart-dio-next/dio_http_petstore_client_lib_fake/pom.xml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                            <goal>exec</goal>
+                        </goals>
+                        <configuration>
+                            <executable>dart</executable>
+                            <arguments>
+                                <argument>test</argument>
+                            </arguments>
+                        </configuration>
+                    </execution>
+                    <execution>
```
Comment (kuhnroyal) [inline samples/openapi3/client/petstore/dart-dio-next/dio_http_petstore_client_lib_fake/pom.xml:84]: Remove this execution

**Review by kuhnroyal — APPROVED** [review]
Nice!

**Comment by lam-nv** [comment]
Hi @josh-burton,

I have question related to dioImport, I'm implement a custom template for my project, with small change, but when config the templateDir in config.yaml then I got the empty value like `import '';` instead `import 'package:dio/dio.dart'`
I have checked by removing the templateDir then I can see the expected result `import 'package:dio/dio.dart'`

config.yaml:
```yaml
generatorName: dart-dio-next
outputDir: ../api_provider
inputSpec: openapi.json
templateDir: src/main/resources/dart/libraries/dio  # my problem is here
typeMappings:
  Client: "ModelClient"
  File: "ModelFile"
  EnumClass: "ModelEnumClass"
additionalProperties:
  hideGenerationTimestamp: "true"
  dioLibrary: dio
  clientName: api_provider
  pubVersion: 1.0.0
  pubName: api_provider
  pubLibrary: api_provider
```
And this is command I used to run the generator:
```bash
openapi-generator-cli generate --config \"./openapi_config.yaml\" --skip-validate-spec 
```
in node environment with package: 
```json
    "@openapitools/openapi-generator-cli": "^2.4.12",
```

Could you advice how to I correct it?

**Pull request merged**